[IV-23-24] Objetivo 9

danieeeld2 opened the pull request:
*Cerrar issues abiertos*:
closes #120  closes #108 closes #105

danieeeld2 committed:
Comienzo creacion de interfaces y simular BD, Avanza #107

Creo structs para tener camiones con matrículas y suministros con IDs. Falat ver como solventar la asignación

danieeeld2 committed:
:construction: Finalizada creación BD. :bug: Fix rutas API closes #111 closes #110 closes #107

Hemos representado las asignaciones dentro de la BD y hemos arreglado las rutas de la API

danieeeld2 committed:
:bug: Fix erratas en bd.go closes #112

danieeeld2 committed:
:bug: Fix erratas en bd.go closes #112

danieeeld2 committed:
:bug: Fix erratas en bd.go closes #112

danieeeld2 committed:
Creacion BD prueba closes #114

danieeeld2 committed:
:construction: Implementada rutas de la API closes #113

Hemos implementado el funcionamiento de las rutas y hemos corregido algún fallo que había de planteamiento en bd.go

danieeeld2 committed:
:construction: Creacion test de la BD avanza #115

Creado test para probar la BD

danieeeld2 committed:
:bug: FIX exportación de campos en JSON closes #116

Para que los campos sean exportables, deben empezar por mayuscula

danieeeld2 committed:
:bug: FIX exportación de campos en JSON closes #116

danieeeld2 committed:
:bug: FIX dependencias indirectas go.mod closes #117

danieeeld2 committed:
:construction: Tests casi implementados avanza #115

Queda pendiente exportar campos de Suministros y Camiones, consultar con JJ

danieeeld2 committed:
:bug: Cambiar mage installdeps por mage install closes #118

danieeeld2 committed:
:bug: Cambiar mage installdeps por mage install closes #118

danieeeld2 commented on internal/suministro.go:
```diff
@@ -3,11 +3,11 @@ package internal
-	peso_kg					float32
-	volumen_cm3				float32
-	tipo					TipoSuministro
+	direccion				string			`json:"direccion"`
```
@JJ está casi acabado, pero me he encontrado con el problema de que para que los parámetros sean exportables, deben estar en mayúscula, no sirve solo con la etiqueta, por lo que tendría que ir por todo el código cambiándolo Voy a esperar a que me digas que tal está el resto y si lo otro va bien, pues ya me pongo a cambiar todo esto (*Esto solo afecta a las rutas `PUT` de `camión` y `suministro`, el resto esta testado*)

JJ replied:
¿Por qué necesitas que los parámetros sean exportables?

danieeeld2 replied:
Como trabajamos con `JSON`, al ser una API REST,  yo, en los tests, creo datos de prueba y los transformo a `JSON` con `jsonResponse, err := json.Marshal(camion)`, pero si los parámetros del struct no están en mayúscula, me dice que no son exportables y da error.

JJ commented on internal/api.go:
```diff
@@ -24,33 +25,174 @@ func createRouter() *chi.Mux{
+
+	jsonResponse, err := json.Marshal(camion)
+	if err != nil {
+		http.Error(w, "Error al convertir a JSON", http.StatusInternalServerError)
```
Os comenté que el `InternalServerError` no era nunca algo que pudiera devolver la aplicación, es algo que devuelve lo que esté delante de la aplicación (proxy, CDN) en caso de que esta no funcione. Piensa qué error podría ser el más adecuado en este caso.

danieeeld2 replied:
Resuelto en commit 228c6d5a6dfd4308bfe46fa69d54de972c2368d9

JJ commented on internal/api_test.go:
```diff
@@ -0,0 +1,338 @@
+		t.Fatal(err)
+	}
+
+	if camion.Matricula != matricula {
```
Igual sería un poco más comprensible que agruparas en subtests.

danieeeld2 replied:
> Igual sería un poco más comprensible que agruparas en subtests.

Okay, lo tendré en cuenta, porque la verdad que hasta yo me estaba liando de tanto código.

danieeeld2 replied:
Resuelto en commit 4f4f1bf2e318117bd5697636905186c6c7116fb6

JJ reviewed (changes requested):
Aparte de lo comentado, no tengo muy claro qué ruta en concreto activa la lógica de negocio. En los tests no está muy claro, más que nada por la cantidad de tests que hay... Quizás sería mejor agruparlos o hacerlo de forma que se vea explícitamente.

danieeeld2 commented:
> Aparte de lo comentado, no tengo muy claro qué ruta en concreto activa la lógica de negocio. En los tests no está muy claro, más que nada por la cantidad de tests que hay... Quizás sería mejor agruparlos o hacerlo de forma que se vea explícitamente.

En sí no la activa ninguna, porque como en la objetivo anterior me pusiste: "igual sería más conveniente un POST que un PUT (donde se supone que ya la tienes, y que la ha calculado el cliente)", pues creí que no debería usarla explicita mente en los test de la `API`. ¿Debería entonces usarla en el test?

JJ commented:
> > Aparte de lo comentado, no tengo muy claro qué ruta en concreto activa la lógica de negocio. En los tests no está muy claro, más que nada por la cantidad de tests que hay... Quizás sería mejor agruparlos o hacerlo de forma que se vea explícitamente.
> 
> En sí no la activa ninguna, porque como en la objetivo anterior me pusiste: "igual sería más conveniente un POST que un PUT (donde se supone que ya la tienes, y que la ha calculado el cliente)", pues creí que no debería usarla explicita mente en los test de la `API`. ¿Debería entonces usarla en el test?

(se pone el gorro de ChatGPT)
Perdón, creo que me expresé mal. Un POST activaría la lógica de negocio. Si usas un PUT, sin embargo, se supone que la lógica de negocio está en otro lado. Por eso te decía que era más conveniente un POST para activar la lógica de negocio.

danieeeld2 commented:
> > > Aparte de lo comentado, no tengo muy claro qué ruta en concreto activa la lógica de negocio. En los tests no está muy claro, más que nada por la cantidad de tests que hay... Quizás sería mejor agruparlos o hacerlo de forma que se vea explícitamente.
> > 
> > 
> > En sí no la activa ninguna, porque como en la objetivo anterior me pusiste: "igual sería más conveniente un POST que un PUT (donde se supone que ya la tienes, y que la ha calculado el cliente)", pues creí que no debería usarla explicita mente en los test de la `API`. ¿Debería entonces usarla en el test?
> 
> (se pone el gorro de ChatGPT) Perdón, creo que me expresé mal. Un POST activaría la lógica de negocio. Si usas un PUT, sin embargo, se supone que la lógica de negocio está en otro lado. Por eso te decía que era más conveniente un POST para activar la lógica de negocio.

😂 buena referencia actual. Okay, lo tendré en cuenta. Voy a ver si entre mañana y pasado lo puedo arreglar, que son los dos últimos días que tengo para dedicarle a la asignatura, por desgracia.

JJ commented:
Ánimo.

danieeeld2 committed:
:construction: El POST de asignacion ahora usa la logica de negocio closes #119

Rediseño del POST /suministro/id/asignacion, para que use la logica de negocio y todos los cambios que ello conlleva

danieeeld2 committed:
:construction: Separar los tests en subtests closes #121

danieeeld2 committed:
:bug: Cambiamos InternalSeverError por BadRequest, closes #122

Siguiendo los consejos de JJ en la revisión, cambiamos el Status

danieeeld2 commented:
> Aparte de lo comentado, no tengo muy claro qué ruta en concreto activa la lógica de negocio. En los tests no está muy claro, más que nada por la cantidad de tests que hay... Quizás sería mejor agruparlos o hacerlo de forma que se vea explícitamente.

Ahora si aplica la lógica de negocio, commit 59da510f0df9feeb9e9552d4a87171239d12d305

danieeeld2 commented:
@JJ he solventado un poco lo más indicativo de la revisión de ayer. Hoy no creo que tenga más tiempo, pero mañana a ver si le puedo sacar un ratillo, por si sigue habiendo errores (que seguramente los haya)

JJ reviewed (approved):
:1st_place_medal: Esencialmente es correcto, aunque podría venirle bien un poco de DRY, evitar DELETEs unnecesarios que no tengan nada que ver con las HUs ni con la lógica de negocio, y agrupar los tests según historias de usuario y no funcionalidades (lo que haría un poco más sencillo su evaluación). Enhorabuena.

danieeeld2 commented:
> :1st_place_medal: Esencialmente es correcto, aunque podría venirle bien un poco de DRY, evitar DELETEs unnecesarios que no tengan nada que ver con las HUs ni con la lógica de negocio, y agrupar los tests según historias de usuario y no funcionalidades (lo que haría un poco más sencillo su evaluación). Enhorabuena.

Parece mentira acabar todos los objetivos😂. Muchas gracias @JJ

danieeeld2 commented:
> > > 🥇 Esencialmente es correcto, aunque podría venirle bien un poco de DRY, evitar DELETEs unnecesarios que no tengan nada que ver con las HUs ni con la lógica de negocio, y agrupar los tests según historias de usuario y no funcionalidades (lo que haría un poco más sencillo su evaluación). Enhorabuena.
> > 
> > 
> > Parece mentira acabar todos los objetivos😂. Muchas gracias @JJ
> 
> El tercero de la historia, si no estoy contando mal... Y el primero fue @amerigal, así que...

Yo creo que este año lo acaba más gente. @Morad02 y @javigp2002 seguro que también lo acaban.

JJ commented:
> > > > 🥇 Esencialmente es correcto, aunque podría venirle bien un poco de DRY, evitar DELETEs unnecesarios que no tengan nada que ver con las HUs ni con la lógica de negocio, y agrupar los tests según historias de usuario y no funcionalidades (lo que haría un poco más sencillo su evaluación). Enhorabuena.
> > > 
> > > 
> > > Parece mentira acabar todos los objetivos😂. Muchas gracias @JJ
> > 
> > 
> > El tercero de la historia, si no estoy contando mal... Y el primero fue @amerigal, así que...
> 
> Yo creo que este año lo acaba más gente. @Morad02 y @javigp2002 seguro que también lo acaban.

También @carlotiii30 está a punto. Pero *no pressure*, ¿eh? ;-)
